src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Tuples.cs

```diff
@@ -0,0 +1,353 @@
+                ValidateTupleNameAttribute(indexer,
+                    expectedTupleNamesAttribute: true,
+                    expectedElementNames: expectedTupleNames);
+                expectedTupleNames = new[] { "e3", "e4" };
+                ValidateTupleNameAttribute(indexer.Parameters.Single(),
+                    expectedTupleNamesAttribute: true,
+                    expectedElementNames: expectedTupleNames);
+            }
+
+            private void ValidateTupleNameAttribute(Symbol symbol,
+                bool expectedTupleNamesAttribute,
+                string[] expectedElementNames = null,
+                bool forReturnType = false)
+            {
+                var synthesizedTupleElementNamesAttr = symbol.GetSynthesizedAttributes(forReturnType)
+                    .Where(attr => string.Equals(attr.AttributeClass.Name,
+                                                 "TupleElementNamesAttribute",
+                                                 StringComparison.Ordinal))
+                    .ToList();
+                if (!expectedTupleNamesAttribute)
+                {
+                    Assert.Empty(synthesizedTupleElementNamesAttr);
+                }
+                else
+                {
+                    var tupleAttr = synthesizedTupleElementNamesAttr.Single();
+                    var expectedCtor = expectedElementNames == null
+                        ? _tupleAttrNoArgs
+                        : _tupleAttrTransformNames;
+                    Assert.NotNull(expectedCtor);
+                    Assert.Equal(expectedCtor, tupleAttr.AttributeConstructor);
+
+                    if (expectedElementNames == null)
+                    {
+                        Assert.Empty(tupleAttr.CommonConstructorArguments);
+                    }
+                    else
+                    {
+                        Assert.Equal(1, tupleAttr.CommonConstructorArguments.Length);
+
+                        var arg = tupleAttr.CommonConstructorArguments[0];
+                        Assert.Equal(TypedConstantKind.Array, arg.Kind);
+
+                        var actualElementNames = arg.Values;
+                        Assert.Equal(expectedElementNames.Length, actualElementNames.Length);
+                        var stringType = _comp.GetSpecialType(SpecialType.System_String);
+
+                        for (int i =  0; i < actualElementNames.Length; i++)
+                        {
+                            string expectedName = expectedElementNames[i];
+                            TypedConstant actualName = actualElementNames[i];
+
+                            Assert.Equal(TypedConstantKind.Primitive, actualName.Kind);
+                            Assert.Equal(stringType, actualName.Type);
+                            Assert.Equal(expectedName, (string)actualName.Value);
+                        }
+                    }
+                }
+            }
+
+        }
+    }
+}
```

Comment:
Some test ideas for now or later:

1. test without String type or maybe a defective one (since I saw you have code pertaining to that)
2. test consuming the metadata (from compilation reference or from image, see `UnifyUnderlyingWithTuple_01`)
3. test loading images where the attribute was placed on other types
4. test with source that explicitly specifies the attribute (should fail)
5. test loading images where the serialized name array doesn't match the type